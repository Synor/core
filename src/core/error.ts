/* eslint-disable no-dupe-class-members */

type MigrationRecord = import('./migration').MigrationRecord

type SynorDatabaseErrorType = 'lock_error' | 'unlock_error'
type SynorMigrationErrorType = 'not_found'
type SynorValidationErrorType = 'dirty' | 'hash_mismatch'

type SynorErrorType =
  | SynorDatabaseErrorType
  | SynorMigrationErrorType
  | SynorValidationErrorType
  | 'EXCEPTION'

export class SynorError extends Error {
  type: SynorErrorType
  meta: any

  constructor(
    message: string,
    meta: any = {},
    type: SynorErrorType = 'EXCEPTION'
  ) {
    super(message)

    Error.captureStackTrace(this, this.constructor)

    this.name = this.constructor.name
    this.type = type
    this.meta = meta
  }
}

type SynorDatabaseErrorMeta = {
  lockId?: string
}

export class SynorDatabaseError extends SynorError {
  type!: SynorDatabaseErrorType
  meta!: SynorDatabaseErrorMeta

  constructor(
    type: 'lock_error' | 'unlock_error',
    meta: Required<Pick<SynorDatabaseErrorMeta, 'lockId'>>
  )

  constructor(type: SynorDatabaseErrorType, meta: SynorDatabaseErrorMeta) {
    super('Database Error', meta, type)
  }
}

export class SynorMigrationError extends SynorError {
  type!: SynorMigrationErrorType
  meta!: Partial<MigrationRecord>

  constructor(type: SynorMigrationErrorType, meta: Partial<MigrationRecord>) {
    super('Migration Error', meta, type)
  }
}

export class SynorValidationError extends SynorError {
  type!: SynorValidationErrorType
  meta!: MigrationRecord

  constructor(type: SynorValidationErrorType, meta: MigrationRecord) {
    super('Validation Error', meta, type)
  }
}
