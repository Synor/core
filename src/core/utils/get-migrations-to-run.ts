import { getMigration } from './get-migration'

type MigrationSource = import('../migration').MigrationSource
type SourceEngine = import('../source').SourceEngine

type Version = MigrationSource['version']
type Type = MigrationSource['type']

export async function getMigrationsToRun(
  source: SourceEngine,
  baseVersion: Version,
  fromVersion: Version,
  toVersion: Version
): Promise<MigrationSource[]> {
  const migrations: MigrationSource[] = []

  let type: Type

  if (fromVersion < toVersion) {
    type = 'DO'
  } else if (fromVersion > toVersion) {
    type = 'UNDO'
  } else {
    return migrations
  }

  if (fromVersion !== baseVersion) {
    const fromVersionExist = await source.get(fromVersion, type)
    if (!fromVersionExist) {
      throw new Error(`NOT_FOUND: Migration(${fromVersion}.${type})`)
    }
  }

  if (toVersion !== baseVersion) {
    const toVersionExist = await source.get(toVersion, type)
    if (!toVersionExist) {
      throw new Error(`NOT_FOUND: Migration(${toVersion}.${type})`)
    }
  }

  if (type === 'DO') {
    let nextVersion =
      fromVersion === baseVersion
        ? await source.first()
        : await source.next(fromVersion)

    while (nextVersion) {
      const migration = await getMigration(source, nextVersion, type)

      if (!migration) {
        nextVersion = null
        break
      }

      migrations.push(migration)

      if (migration.version === toVersion) {
        nextVersion = null
        break
      }

      nextVersion = await source.next(migration.version)
    }
  }

  if (type === 'UNDO') {
    let prevVersion = fromVersion === baseVersion ? null : fromVersion

    while (prevVersion) {
      const migration = await getMigration(source, prevVersion, type)

      if (!migration) {
        prevVersion = null
        break
      }

      if (migration.version === toVersion) {
        prevVersion = null
        break
      }

      migrations.push(migration)

      prevVersion = await source.prev(migration.version)
    }
  }

  return migrations
}
