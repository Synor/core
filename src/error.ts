/* eslint-disable no-dupe-class-members */

type MigrationRecord = import('./migration').MigrationRecord

type SynorMigrationErrorType = 'not_found'
type SynorValidationErrorType = 'dirty' | 'hash_mismatch'

type SynorErrorType =
  | SynorMigrationErrorType
  | SynorValidationErrorType
  | 'invalid_filename'
  | 'exception'

type SynorErrorDataType = {
  dirty: MigrationRecord
  hash_mismatch: MigrationRecord
  not_found: Partial<MigrationRecord>
  invalid_filename: { filename: string }
  exception: Error | Record<string, any>
}

export class SynorError<
  ErrorType extends SynorErrorType = 'exception'
> extends Error {
  type: ErrorType
  data: SynorErrorDataType[ErrorType]

  /**
   * Creates a general instance of SynorError
   *
   * @param message Error message
   */
  constructor(message: string)

  /**
   * Creates a special instance of SynorError with `type` and `data`
   *
   * @param message Error message
   * @param type Error type
   * @param data Error data
   */
  constructor(
    message: string,
    type: ErrorType,
    data: SynorErrorDataType[ErrorType]
  )

  constructor(
    message: string,
    type: SynorErrorType = 'exception',
    data: SynorErrorDataType[SynorErrorType] = {}
  ) {
    super(message)

    Error.captureStackTrace(this, this.constructor)

    if (data instanceof Error) {
      this.stack = `${data.stack}\n${this.stack}`
    }

    this.name = this.constructor.name
    this.type = type as ErrorType
    this.data = data as SynorErrorDataType[ErrorType]
  }
}

export function toSynorError<T extends SynorErrorType = 'exception'>(
  error: Error | SynorError<T>
): SynorError<T> {
  if (error instanceof SynorError) {
    return error
  }

  return new SynorError(
    `Exception: ${error.message}`,
    'exception',
    error
  ) as SynorError<T>
}
