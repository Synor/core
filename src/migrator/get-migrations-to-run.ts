import { SynorError } from '../error'
import { getCurrentRecord } from './get-current-record'
import { getMigration } from './get-migration'

type MigrationRecordInfo = import('../migration').MigrationRecordInfo
type MigrationSourceInfo = import('../migration').MigrationSourceInfo
type MigrationType = import('../migration').MigrationType
type SourceEngine = import('../source').SourceEngine

export async function getMigrationsToRun({
  recordInfos,
  source,
  baseVersion,
  targetVersion: toVersion,
  outOfOrder
}: {
  recordInfos: MigrationRecordInfo[]
  source: SourceEngine
  baseVersion: string
  targetVersion: string
  outOfOrder: boolean
}): Promise<MigrationSourceInfo[]> {
  const appliedVersions = recordInfos
    .filter(({ state, type }) => state === 'applied' && type === 'do')
    .map(({ version }) => version)

  const fromVersion = getCurrentRecord(recordInfos).version

  const migrations: MigrationSourceInfo[] = []

  let type: MigrationType

  if (fromVersion < baseVersion) {
    throw new SynorError(
      `fromVersion(${fromVersion}) is below baseVersion(${baseVersion})`
    )
  }

  if (toVersion < baseVersion) {
    throw new SynorError(
      `toVersion(${toVersion}) is below baseVersion(${baseVersion})`
    )
  }

  if (fromVersion < toVersion) {
    type = 'do'
  } else if (fromVersion > toVersion) {
    type = 'undo'
  } else if (outOfOrder) {
    type = 'do'
  } else {
    return migrations
  }

  if (fromVersion !== baseVersion) {
    const fromVersionExist = await source.get(fromVersion, type)
    if (!fromVersionExist) {
      throw new SynorError(
        `Missing Migration Source => Version(${fromVersion}) Type(${type})`,
        'not_found',
        { version: fromVersion, type }
      )
    }
  }

  if (toVersion !== baseVersion) {
    const toVersionExist = await source.get(toVersion, type)
    if (!toVersionExist) {
      throw new SynorError(
        `Missing Migration Source => Version(${toVersion}) Type(${type})`,
        'not_found',
        { version: toVersion, type }
      )
    }
  }

  if (type === 'do') {
    const startVersion = outOfOrder ? baseVersion : fromVersion

    let nextVersion = await source.next(startVersion)

    if (!nextVersion && startVersion === baseVersion) {
      nextVersion = await source.first()
    }

    while (nextVersion) {
      const migration = await getMigration(source, nextVersion, type)

      if (!migration) {
        nextVersion = null
        break
      }

      if (!appliedVersions.includes(migration.version)) {
        migrations.push({ ...migration, state: 'pending' })
      }

      if (migration.version === toVersion) {
        nextVersion = null
        break
      }

      nextVersion = await source.next(migration.version)
    }
  }

  if (type === 'undo') {
    let currVersion = fromVersion === baseVersion ? null : fromVersion

    while (currVersion) {
      const migration = await getMigration(source, currVersion, type)

      if (!migration) {
        currVersion = null
        break
      }

      if (migration.version === toVersion) {
        currVersion = null
        break
      }

      if (appliedVersions.includes(migration.version)) {
        migrations.push({ ...migration, state: 'pending' })
      }

      currVersion = await source.prev(migration.version)
    }
  }

  return migrations
}
