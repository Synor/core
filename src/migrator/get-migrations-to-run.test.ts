import { SynorError } from '../error'
import { getMigrationInfo } from '../__utils__.test'
import { getMigrationsToRun } from './get-migrations-to-run'

type MigrationInfo = ReturnType<typeof getMigrationInfo>
type MigrationSource = import('../index').MigrationSource
type SourceEngine = import('../index').SourceEngine

const infoMap: Record<
  MigrationInfo['version'],
  Record<MigrationInfo['type'], MigrationInfo | undefined>
> = {
  '02': {
    do: getMigrationInfo({ version: '02', type: 'do' }),
    undo: getMigrationInfo({ version: '02', type: 'undo' })
  },
  '03': {
    do: getMigrationInfo({ version: '03', type: 'do' }),
    undo: getMigrationInfo({ version: '03', type: 'undo' })
  },
  '04': {
    do: getMigrationInfo({ version: '04', type: 'do' }),
    undo: getMigrationInfo({ version: '04', type: 'undo' })
  },
  '05': {
    do: getMigrationInfo({ version: '05', type: 'do' }),
    undo: undefined
  },
  '06': {
    do: undefined,
    undo: getMigrationInfo({ version: '06', type: 'undo' })
  },
  '07': {
    do: getMigrationInfo({ version: '07', type: 'do' }),
    undo: getMigrationInfo({ version: '07', type: 'undo' })
  }
}
const versions = Object.keys(infoMap)

const formatResult = (items: MigrationSource[]): string[] => {
  return items.map(({ version, type }) => `${version}.${type}`)
}

describe('migrator:getMigrationsToRun', () => {
  const source: SourceEngine = {
    open: () => Promise.resolve(),
    close: () => Promise.resolve(),
    first: () => Promise.resolve(versions[0]),
    prev: (v: MigrationInfo['version']) => {
      const currentIndex = versions.indexOf(v)
      return Promise.resolve(
        currentIndex === -1 ? null : versions[currentIndex - 1] || null
      )
    },
    next: (v: MigrationInfo['version']) => {
      const currentIndex = versions.indexOf(v)
      return Promise.resolve(
        currentIndex === -1 ? null : versions[currentIndex + 1] || null
      )
    },
    last: () => Promise.resolve(versions[versions.length - 1]),
    get: (version: MigrationInfo['version'], type: MigrationInfo['type']) => {
      const m = infoMap[version]
      return Promise.resolve(m ? m[type] || null : null)
    },
    read: () => Promise.resolve(Buffer.from(''))
  }

  test('works (from=to)', async () => {
    const result = await getMigrationsToRun(source as any, '0', '01', '01')
    expect(formatResult(result)).toMatchInlineSnapshot(`Array []`)
  })

  test('#23 throws if (from<base)', async () => {
    await expect(
      getMigrationsToRun(source as any, '02', '01', '03')
    ).rejects.toMatchInlineSnapshot(
      `[SynorError: fromVersion(01) is below baseVersion(02)]`
    )
  })

  test('#23 throws if (to<base)', async () => {
    await expect(
      getMigrationsToRun(source as any, '02', '03', '01')
    ).rejects.toMatchInlineSnapshot(
      `[SynorError: toVersion(01) is below baseVersion(02)]`
    )
  })

  test('throws (from<to ; from not exists)', async () => {
    try {
      await getMigrationsToRun(source as any, '0', '01', '05')
    } catch (error) {
      expect(error).toBeInstanceOf(SynorError)
      expect(error.data).toMatchInlineSnapshot(`
        Object {
          "type": "do",
          "version": "01",
        }
      `)
      expect(error.type).toMatchInlineSnapshot(`"not_found"`)
    }
  })

  test('throws (from<to ; to not exists)', async () => {
    try {
      await getMigrationsToRun(source as any, '0', '02', '99')
    } catch (error) {
      expect(error).toBeInstanceOf(SynorError)
      expect(error.data).toMatchInlineSnapshot(`
        Object {
          "type": "do",
          "version": "99",
        }
      `)
      expect(error.type).toMatchInlineSnapshot(`"not_found"`)
    }
  })

  test('works (from<to)', async () => {
    const result = await getMigrationsToRun(source as any, '0', '02', '04')
    expect(formatResult(result)).toMatchInlineSnapshot(`
      Array [
        "03.do",
        "04.do",
      ]
    `)
  })

  test('works (from<to ; from is base)', async () => {
    const result = await getMigrationsToRun(source as any, '01', '01', '04')
    expect(formatResult(result)).toMatchInlineSnapshot(`
      Array [
        "02.do",
        "03.do",
        "04.do",
      ]
    `)
  })

  test('works (from<to ; not exists in middle)', async () => {
    const result = await getMigrationsToRun(source as any, '0', '02', '07')
    expect(formatResult(result)).toMatchInlineSnapshot(`
      Array [
        "03.do",
        "04.do",
        "05.do",
      ]
    `)
  })

  test('throws (from>to ; from not exists)', async () => {
    try {
      await getMigrationsToRun(source as any, '0', '99', '02')
    } catch (error) {
      expect(error).toBeInstanceOf(SynorError)
      expect(error.data).toMatchInlineSnapshot(`
        Object {
          "type": "undo",
          "version": "99",
        }
      `)
      expect(error.type).toMatchInlineSnapshot(`"not_found"`)
    }
  })

  test('throws (from>to ; to not exists)', async () => {
    try {
      await getMigrationsToRun(source as any, '0', '04', '01')
    } catch (error) {
      expect(error).toBeInstanceOf(SynorError)
      expect(error.data).toMatchInlineSnapshot(`
        Object {
          "type": "undo",
          "version": "01",
        }
      `)
      expect(error.type).toMatchInlineSnapshot(`"not_found"`)
    }
  })

  test('works (from>to)', async () => {
    const result = await getMigrationsToRun(source as any, '0', '04', '02')
    expect(formatResult(result)).toMatchInlineSnapshot(`
      Array [
        "04.undo",
        "03.undo",
      ]
    `)
  })

  test('#23 throws (from>to ; from is base)', async () => {
    await expect(
      getMigrationsToRun(source as any, '03', '03', '02')
    ).rejects.toMatchInlineSnapshot(
      `[SynorError: toVersion(02) is below baseVersion(03)]`
    )
  })

  test('works (from>to ; to is base)', async () => {
    const result = await getMigrationsToRun(source as any, '01', '04', '01')
    expect(formatResult(result)).toMatchInlineSnapshot(`
      Array [
        "04.undo",
        "03.undo",
        "02.undo",
      ]
    `)
  })

  test('works (from>to ; not exists in middle)', async () => {
    const result = await getMigrationsToRun(source as any, '0', '07', '02')
    expect(formatResult(result)).toMatchInlineSnapshot(`
      Array [
        "07.undo",
        "06.undo",
      ]
    `)
  })

  test('#22 works (from<to ; from=base ; source: first<base<last)', async () => {
    const result = await getMigrationsToRun(source as any, '03', '03', '05')
    expect(formatResult(result)).toMatchInlineSnapshot(`
      Array [
        "04.do",
        "05.do",
      ]
    `)
  })
})
