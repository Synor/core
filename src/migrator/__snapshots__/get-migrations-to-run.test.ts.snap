// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrator:getMigrationsToRun works (from<to ; from is base) 1`] = `
Array [
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "02",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "03",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "04",
  },
]
`;

exports[`migrator:getMigrationsToRun works (from<to ; not exists in middle) 1`] = `
Array [
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "03",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "04",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "05",
  },
]
`;

exports[`migrator:getMigrationsToRun works (from<to) 1`] = `
Array [
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "03",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "do",
    "version": "04",
  },
]
`;

exports[`migrator:getMigrationsToRun works (from=to) 1`] = `Array []`;

exports[`migrator:getMigrationsToRun works (from>to ; from is base) 1`] = `Array []`;

exports[`migrator:getMigrationsToRun works (from>to ; not exists in middle) 1`] = `
Array [
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "undo",
    "version": "07",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "undo",
    "version": "06",
  },
]
`;

exports[`migrator:getMigrationsToRun works (from>to ; to is base) 1`] = `
Array [
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "undo",
    "version": "04",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "undo",
    "version": "03",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "undo",
    "version": "02",
  },
]
`;

exports[`migrator:getMigrationsToRun works (from>to) 1`] = `
Array [
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "undo",
    "version": "04",
  },
  Object {
    "body": "",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "title": "",
    "type": "undo",
    "version": "03",
  },
]
`;
